File: 5회차/14_longest_common_prefix/14_longest_common_prefix.cpp
#include <iostream>
#include <string>
#include <vector>

// Function to find the longest common prefix
std::string longest_common_prefix(const std::vector<std::string>& strs) {
    if (strs.empty()) {
        return "";
    }

    // Start with the first string as the prefix
    std::string prefix = strs[0];

    // Iterate over the remaining strings
    for (size_t i = 1; i < strs.size(); i++) {
        // Compare prefix with the current string
        while (strs[i].find(prefix) != 0) {
            prefix.pop_back(); // Remove the last character from prefix
            if (prefix.empty()) {
                return ""; // No common prefix
            }
        }
    }

    return prefix;
}

File: 5회차/14_longest_common_prefix/14_longest_common_prefix.py
def longest_common_prefix(strs): # input: list of words
    if not strs: # 아무것도 없을 때
        return "" # 빈 string return
    
    # 일단 첫번째 단어를 (통으로) prefix 로 지정
    prefix = strs[0]
    
    # prefix와, 나머지 단어들 순서대로 하나씩 비교
    for string in strs[1:]:
        # prefix 와 같지 않으면
        while not string.startswith(prefix):
            prefix = prefix[:-1] # prefix를 끝 알파벳에서부터 줄여나감
            if not prefix: # prefix 가 empty 되면
                return ""  # return 빈 string
    
    return prefix